File: app/src/main/kotlin/uk/co/twinscrollgridbalancer/tsgbheater/ble/FrameCodec.kt
package uk.co.twinscrollgridbalancer.tsgbheater.ble

import kotlin.random.Random

// Frame builder/parser for HeatGenie's BLE protocol. See
// docs/BLE_PROTOCOL.md for line-by-line citations into the decompiled JS.
//
// Wire format (8-byte heater frame):
//   AA  len  op  d0  d1  d2  crcH  crcL    (CRC-16/XMODEM over bytes 0..5)
//
// Telemetry response variant (longer):
//   AA  len  op  d0  d1  d2  d3  tag  <payload, 4*(len+1) bytes>  crcH crcL
//
// Where tag at byte 7 selects an "info area":
//   0xF1 = version, 0xF2 = regInfo (live status), 0xF3 = paraInfo,
//   0xF5 = timerInfo.
object FrameCodec {

    private const val SYNC: Byte = 0xAA.toByte()

    // Heater-side opcodes (DB0_DN_*).
    private const val OP_CMD             = 0x61   // button events (uses random nonce in d1)
    private const val OP_AUTO_UPDATE     = 0x65   // start/stop periodic telemetry
    private const val OP_SHORT_PARA      = 0x66   // set a single setting
    private const val OP_GET_REG_VAL     = 0x64   // one-shot read of an info area

    // CMD_* values used as d0 of DB0_DN_CMD.
    private const val CMD_ON              = 1
    private const val CMD_OFF             = 2
    private const val CMD_UP              = 3
    private const val CMD_DOWN            = 4
    private const val CMD_SWITCH_TEMP_FC  = 8
    private const val CMD_BLOW_ON         = 9
    private const val CMD_SWITCH_TEMP_CF  = 10
    private const val CMD_OIL_PUMP_ON     = 11
    private const val CMD_OIL_PUMP_OFF    = 12

    // SHORT_* values used as d0 of DB0_DN_SHORT_PARA.
    private const val SHORT_RUN_MODE      = 0x00
    private const val SHORT_TARGET_TEMP   = 0x01
    private const val SHORT_TARGET_GEAR   = 0x02
    private const val SHORT_TEMP_DIFF     = 0x04

    // Area tags.
    const val AREA_VERSION  = 0xF1
    const val AREA_REG_INFO = 0xF2
    const val AREA_PARA     = 0xF3
    const val AREA_TIMER    = 0xF5

    // Temperature units used as d1 of SHORT_TARGET_TEMP / SHORT_TEMP_DIFF.
    enum class TempUnit(val wire: Int) { Celsius(0), Fahrenheit(1) }

    enum class RunMode(val wire: Int) { Auto(0), Manual(1), StartStop(2) }

    // --- Outgoing command builders -------------------------------------

    fun buildStartHeater(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_ON.toByte(),  randomNonce(), 0))

    fun buildStopHeater(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_OFF.toByte(), randomNonce(), 0))

    fun buildButtonUp(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_UP.toByte(),   randomNonce(), 0))

    fun buildButtonDown(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_DOWN.toByte(), randomNonce(), 0))

    fun buildBlowOn(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_BLOW_ON.toByte(),     randomNonce(), 0))

    fun buildOilPumpOn(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_OIL_PUMP_ON.toByte(), randomNonce(), 0))

    fun buildOilPumpOff(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_OIL_PUMP_OFF.toByte(), randomNonce(), 0))

    fun buildSwitchToCelsius(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_SWITCH_TEMP_FC.toByte(), randomNonce(), 0))

    fun buildSwitchToFahrenheit(): ByteArray =
        eightByteCmd(OP_CMD, byteArrayOf(CMD_SWITCH_TEMP_CF.toByte(), randomNonce(), 0))

    fun buildSetTargetTemp(value: Int, unit: TempUnit = TempUnit.Celsius): ByteArray =
        eightByteCmd(OP_SHORT_PARA, byteArrayOf(
            SHORT_TARGET_TEMP.toByte(),
            unit.wire.toByte(),
            (value and 0xFF).toByte(),
        ))

    fun buildSetGear(gear: Int): ByteArray =
        eightByteCmd(OP_SHORT_PARA, byteArrayOf(
            SHORT_TARGET_GEAR.toByte(),
            0,
            (gear and 0xFF).toByte(),
        ))

    fun buildSetRunMode(mode: RunMode): ByteArray =
        eightByteCmd(OP_SHORT_PARA, byteArrayOf(
            SHORT_RUN_MODE.toByte(),
            0,
            mode.wire.toByte(),
        ))

    fun buildSetTempHysteresis(diff: Int, unit: TempUnit = TempUnit.Celsius): ByteArray =
        eightByteCmd(OP_SHORT_PARA, byteArrayOf(
            SHORT_TEMP_DIFF.toByte(),
            unit.wire.toByte(),
            (diff and 0xFF).toByte(),
        ))

    // Start the periodic telemetry stream. The vendor app sends (0x0A, 0x00)
    // for ~1 Hz when a page opens; we follow that convention.
    fun buildStartTelemetryStream(): ByteArray =
        eightByteCmd(OP_AUTO_UPDATE, byteArrayOf(0x02, 0x0A, 0x00))

    fun buildStopTelemetryStream(): ByteArray =
        eightByteCmd(OP_AUTO_UPDATE, byteArrayOf(0x02, 0x00, 0x1E))

    fun buildReadRegInfo(): ByteArray =
        eightByteCmd(OP_GET_REG_VAL, byteArrayOf(AREA_REG_INFO.toByte(), 0x00, 0x05))

    // Read the 7-day timer schedule (area 0xF5). Reply arrives as the longer
    // telemetry response variant with tag 0xF5; pass through parseTimerInfo.
    fun buildReadTimerInfo(): ByteArray =
        eightByteCmd(OP_GET_REG_VAL, byteArrayOf(AREA_TIMER.toByte(), 0x00, 0x09))

    private fun eightByteCmd(opcode: Int, data: ByteArray): ByteArray {
        require(data.size == 3) { "Command payload must be exactly 3 bytes" }
        val out = ByteArray(8)
        out[0] = SYNC
        out[1] = 0x00
        out[2] = opcode.toByte()
        data.copyInto(out, destinationOffset = 3)
        val crc = crc16Xmodem(out, 0, 6)
        out[6] = ((crc shr 8) and 0xFF).toByte()
        out[7] = (crc and 0xFF).toByte()
        return out
    }

    private fun randomNonce(): Byte = Random.nextInt(0, 256).toByte()

    // --- CRC-16/XMODEM -------------------------------------------------
    // Poly 0x1021, init 0x0000, no reflection, no final XOR. Table lookup
    // mirrors the JS implementation in HeatGenie's Ee() function.

    private val CRC_TABLE: IntArray = intArrayOf(
        0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50A5, 0x60C6, 0x70E7,
        0x8108, 0x9129, 0xA14A, 0xB16B, 0xC18C, 0xD1AD, 0xE1CE, 0xF1EF,
    )

    private fun crc16Xmodem(buf: ByteArray, offset: Int, len: Int): Int {
        var crc = 0
        for (i in 0 until len) {
            val b = buf[offset + i].toInt() and 0xFF
            // Two nibble lookups per byte, mirroring the JS table size (16).
            val hi = (b shr 4) and 0x0F
            val lo = b and 0x0F
            crc = ((crc shl 4) and 0xFFFF) xor CRC_TABLE[((crc shr 12) and 0x0F) xor hi]
            crc = ((crc shl 4) and 0xFFFF) xor CRC_TABLE[((crc shr 12) and 0x0F) xor lo]
        }
        return crc and 0xFFFF
    }

    fun validateCrc(bytes: ByteArray): Boolean {
        if (bytes.size < 4) return false
        // Per the JS: crc16(buf, total_len) == 0 for a valid frame.
        return crc16Xmodem(bytes, 0, bytes.size) == 0
    }

    // --- Incoming frames -----------------------------------------------

    fun parseTelemetry(bytes: ByteArray): HeaterTelemetry? {
        // Reject anything that isn't a heater frame or has a bad CRC.
        if (bytes.size < 10) return null
        if (bytes[0] != SYNC) return null
        if (!validateCrc(bytes)) return null

        val len = bytes[1].toInt() and 0xFF
        if (len == 0) return null // 8-byte short frames carry no telemetry payload

        val tag = bytes[7].toInt() and 0xFF
        if (tag != AREA_REG_INFO) return null // only regInfo is telemetry-rich

        // Payload starts at byte 8, runs 4*(len+1) bytes.
        val payloadLen = 4 * (len + 1)
        if (bytes.size < 8 + payloadLen + 2) return null
        val p = bytes.copyOfRange(8, 8 + payloadLen)
        if (p.size < 24) return null // need at least the basic block

        val machineStatus  = p[0].toInt() and 0xFF
        val runningModeNib = machineStatus and 0x0F
        val tempUnitF      = (machineStatus and 0x10) != 0    // bit 4 = °F flag

        val voltage   = uleShort(p, 2) / 10.0
        val altitude  = uleShort(p, 4)
        val ambient   = signedTempC(p, 6)
        val housing   = signedTempC(p, 8)
        val pumpHz    = uleShort(p, 10) / 10.0
        val ignWatts  = uleShort(p, 12) / 10.0
        val fanRpm    = uleShort(p, 14)
        val intake    = signedTempC(p, 16).takeUnless { it == NOT_SUPPORTED }
        val outlet    = signedTempC(p, 18).takeUnless { it == NOT_SUPPORTED }
        val errorBits = uleShort(p, 20)

        val aimTarget = if (p.size >= 36) {
            val gear = p[32].toInt() and 0xFF
            val tval = p[33].toInt() and 0xFF
            // The byte holds the target in the heater's CURRENT display unit
            // (tempUnitFlag, bit 4 of machineStatus). Convert F→C so the app
            // always works in Celsius internally; HeaterTelemetry.targetTempC
            // is canonical Celsius regardless of what the heater is showing.
            val targetC = if (tempUnitF) (tval - 32) * 5.0 / 9.0 else tval.toDouble()
            AimTarget(gear = gear, valueC = targetC)
        } else null

        return HeaterTelemetry(
            outletTempC      = outlet,
            targetTempC      = aimTarget?.valueC,
            fuelPumpHz       = pumpHz,
            fanRpm           = fanRpm,
            glowPlugA        = null,
            batteryV         = voltage,
            ambientTempC     = ambient,
            housingTempC     = housing,
            intakeTempC      = intake,
            altitudeM        = altitude,
            ignitionWatts    = ignWatts,
            runningMode      = RunningMode.fromWire(runningModeNib),
            faultBits        = errorBits,
            tempUnitFahrenheit = tempUnitF,
            aimGear          = aimTarget?.gear,
            updatedAtMs      = System.currentTimeMillis(),
        )
    }

    // Parse the timerInfoArea (tag 0xF5) reply into 7 weekday slots.
    // Returns null if the frame doesn't match expectations.
    fun parseTimerInfo(bytes: ByteArray): List<TimerSlot>? {
        if (bytes.size < 10) return null
        if (bytes[0] != SYNC) return null
        if (!validateCrc(bytes)) return null
        val len = bytes[1].toInt() and 0xFF
        if (len == 0) return null
        if ((bytes[7].toInt() and 0xFF) != AREA_TIMER) return null
        val payloadLen = 4 * (len + 1)
        if (bytes.size < 8 + payloadLen + 2) return null
        if (payloadLen < 35) return null   // need 7 × 5 bytes minimum
        val p = bytes.copyOfRange(8, 8 + payloadLen)
        return List(7) { i ->
            val off = i * 5
            val rawMode = p[off + 0].toInt() and 0xFF
            // Vendor app folds raw 1/2 to display value 3 and ≥4 to 4.
            val folded = when {
                rawMode == 1 || rawMode == 2 -> 3
                rawMode >= 4                 -> 4
                else                         -> rawMode
            }
            TimerSlot(
                dayIndex = i,
                modeRaw  = rawMode,
                mode     = TimerMode.fromFolded(folded),
                onHour   = p[off + 1].toInt() and 0xFF,
                onMin    = p[off + 2].toInt() and 0xFF,
                offHour  = p[off + 3].toInt() and 0xFF,
                offMin   = p[off + 4].toInt() and 0xFF,
            )
        }
    }

    private fun uleShort(p: ByteArray, off: Int): Int =
        ((p[off + 1].toInt() and 0xFF) shl 8) or (p[off].toInt() and 0xFF)

    private fun signedTempC(p: ByteArray, off: Int): Double {
        val raw = uleShort(p, off)
        if (raw == 32760) return NOT_SUPPORTED   // sentinel
        val signed = if (raw > 32767) raw - 65536 else raw
        return signed / 10.0
    }

    private const val NOT_SUPPORTED = Double.MIN_VALUE

    // --- Hex helpers ----------------------------------------------------

    fun hex(bytes: ByteArray): String =
        bytes.joinToString(separator = " ") { "%02X".format(it) }

    fun fromHex(hex: String): ByteArray {
        val cleaned = hex.filter { !it.isWhitespace() && it != ':' && it != ',' }
        require(cleaned.length % 2 == 0) { "Hex must have an even length" }
        return ByteArray(cleaned.length / 2) { i ->
            cleaned.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }

    private data class AimTarget(val gear: Int, val valueC: Double)
}

// Running-mode nibble (low 4 bits of machineStatus byte). String labels
// match the vendor app's UI.
enum class RunningMode(val wire: Int, val label: String) {
    Boot(0,             "Boot"),
    Ignition(1,         "Ignition"),
    AutoRun(2,          "Running"),
    ManualRun(3,        "Manual"),
    Cooldown(4,         "Cool-down"),
    Standby(5,          "Standby"),
    Fault(6,            "Fault"),
    ManualPump(7,       "Priming pump"),
    Ventilation(8,      "Ventilating"),
    StartStopActive(9,  "Auto on"),
    StartStopConfig(10, "Auto configured"),
    Unknown(-1,         "—");

    companion object {
        fun fromWire(v: Int): RunningMode = entries.firstOrNull { it.wire == v } ?: Unknown
    }
}

File: windows/src/Ember/Ble/FrameCodec.cs
namespace Ember.Ble;

// Frame builder/parser for the HeatGenie BLE protocol. Line-for-line
// port of android/app/.../ble/FrameCodec.kt. See docs/BLE_PROTOCOL.md for
// citations into the decompiled vendor app.
//
// Wire format (8-byte heater frame):
//   AA  len  op  d0  d1  d2  crcH  crcL    (CRC-16/XMODEM over bytes 0..5)
//
// Telemetry response variant (longer):
//   AA  len  op  d0  d1  d2  d3  tag  <payload, 4*(len+1) bytes>  crcH crcL
//
// Long-write variant (paraInfo / timerInfo writes):
//   AA <len> FF <rand> 00 00 00 <areaTag> <payload> <payloadCrc> <frameCrc>
//   payload must be 4-aligned; len = (payloadBytes/4) - 1
public static class FrameCodec
{
    private const byte SYNC = 0xAA;

    // Heater-side opcodes (DB0_DN_*)
    private const int OP_CMD          = 0x61;   // button events (random nonce in d1)
    private const int OP_MANU_PUMP    = 0x62;   // timed manual pump (seconds in d0)
    private const int OP_AUTO_UPDATE  = 0x65;   // start/stop periodic telemetry
    private const int OP_SHORT_PARA   = 0x66;   // set a single setting
    private const int OP_GET_REG_VAL  = 0x64;   // one-shot read of an info area
    private const int OP_WRITE_AREA   = 0xFF;   // long-frame write to a whole info area

    // CMD_* (d0 of DB0_DN_CMD)
    private const int CMD_ON             = 1;
    private const int CMD_OFF            = 2;
    private const int CMD_UP             = 3;
    private const int CMD_DOWN           = 4;
    private const int CMD_SWITCH_TEMP_FC = 8;   // switch panel to °C (was in °F)
    private const int CMD_BLOW_ON        = 9;
    private const int CMD_SWITCH_TEMP_CF = 10;  // switch panel to °F (was in °C)
    private const int CMD_OIL_PUMP_ON    = 11;
    private const int CMD_OIL_PUMP_OFF   = 12;  // NB: vendor's manual-pump-off path uses CMD_OFF instead.

    // SHORT_* (d0 of DB0_DN_SHORT_PARA)
    private const int SHORT_RUN_MODE     = 0x00;
    private const int SHORT_TARGET_TEMP  = 0x01;
    private const int SHORT_TARGET_GEAR  = 0x02;
    private const int SHORT_TIMER_ADJUST = 0x03;   // also overloaded as wall-clock setter
    private const int SHORT_TEMP_DIFF    = 0x04;

    // Area tags
    public const int AREA_VERSION  = 0xF1;
    public const int AREA_REG_INFO = 0xF2;
    public const int AREA_PARA     = 0xF3;
    public const int AREA_TIMER    = 0xF5;

    public enum TempUnit { Celsius = 0, Fahrenheit = 1 }
    public enum RunMode  { Auto    = 0, Manual     = 1, StartStop = 2 }

    public const int ManualPumpMinSeconds     = 5;
    public const int ManualPumpMaxSeconds     = 90;
    public const int ManualPumpDefaultSeconds = 20;

    // --- 8-byte command builders ---------------------------------------

    public static byte[] BuildStartHeater() =>
        Eight(OP_CMD, (byte)CMD_ON,  Nonce(), 0);

    public static byte[] BuildStopHeater() =>
        Eight(OP_CMD, (byte)CMD_OFF, Nonce(), 0);

    public static byte[] BuildButtonUp() =>
        Eight(OP_CMD, (byte)CMD_UP,   Nonce(), 0);

    public static byte[] BuildButtonDown() =>
        Eight(OP_CMD, (byte)CMD_DOWN, Nonce(), 0);

    public static byte[] BuildBlowOn() =>
        Eight(OP_CMD, (byte)CMD_BLOW_ON,      Nonce(), 0);

    public static byte[] BuildOilPumpOnButton() =>
        Eight(OP_CMD, (byte)CMD_OIL_PUMP_ON,  Nonce(), 0);

    public static byte[] BuildOilPumpOffButton() =>
        Eight(OP_CMD, (byte)CMD_OIL_PUMP_OFF, Nonce(), 0);

    // Timed manual pump. The CMD_OIL_PUMP_ON button alone won't spin the
    // pump — the controller waits for this opcode with a duration.
    public static byte[] BuildManualPumpRun(int seconds)
    {
        int s = Math.Clamp(seconds, ManualPumpMinSeconds, ManualPumpMaxSeconds);
        return Eight(OP_MANU_PUMP, (byte)s, Nonce(), 0);
    }

    public static byte[] BuildSwitchToCelsius() =>
        Eight(OP_CMD, (byte)CMD_SWITCH_TEMP_FC, Nonce(), 0);

    public static byte[] BuildSwitchToFahrenheit() =>
        Eight(OP_CMD, (byte)CMD_SWITCH_TEMP_CF, Nonce(), 0);

    public static byte[] BuildSetTargetTemp(int value, TempUnit unit = TempUnit.Celsius) =>
        Eight(OP_SHORT_PARA,
            (byte)SHORT_TARGET_TEMP,
            (byte)(int)unit,
            (byte)(value & 0xFF));

    public static byte[] BuildSetGear(int gear) =>
        Eight(OP_SHORT_PARA,
            (byte)SHORT_TARGET_GEAR,
            0,
            (byte)(gear & 0xFF));

    public static byte[] BuildSetRunMode(RunMode mode) =>
        Eight(OP_SHORT_PARA,
            (byte)SHORT_RUN_MODE,
            0,
            (byte)(int)mode);

    public static byte[] BuildSetTempHysteresis(int diff, TempUnit unit = TempUnit.Celsius) =>
        Eight(OP_SHORT_PARA,
            (byte)SHORT_TEMP_DIFF,
            (byte)(int)unit,
            (byte)(diff & 0xFF));

    // Set the heater's wall clock. Vendor convention: Monday=0..Sunday=6.
    // hour is 5 bits (0..31), minute 0..59.
    public static byte[] BuildSetClock(int vendorDayMonZero, int hour, int minute)
    {
        int day = Math.Clamp(vendorDayMonZero, 0, 6);
        int h   = Math.Clamp(hour,   0, 31);
        int m   = Math.Clamp(minute, 0, 59);
        int packed = (day << 5) | h;
        return Eight(OP_SHORT_PARA,
            (byte)SHORT_TIMER_ADJUST,
            (byte)packed,
            (byte)m);
    }

    // Translate a DayOfWeek (Sunday=0..Saturday=6) into the vendor's
    // Mon=0..Sun=6 index.
    public static int DayOfWeekToVendor(DayOfWeek dow) => dow switch
    {
        DayOfWeek.Monday    => 0,
        DayOfWeek.Tuesday   => 1,
        DayOfWeek.Wednesday => 2,
        DayOfWeek.Thursday  => 3,
        DayOfWeek.Friday    => 4,
        DayOfWeek.Saturday  => 5,
        DayOfWeek.Sunday    => 6,
        _                    => 0,
    };

    public static byte[] BuildStartTelemetryStream() =>
        Eight(OP_AUTO_UPDATE, 0x02, 0x0A, 0x00);

    public static byte[] BuildStopTelemetryStream() =>
        Eight(OP_AUTO_UPDATE, 0x02, 0x00, 0x1E);

    public static byte[] BuildReadRegInfo() =>
        Eight(OP_GET_REG_VAL, (byte)AREA_REG_INFO, 0x00, 0x05);

    public static byte[] BuildReadTimerInfo() =>
        Eight(OP_GET_REG_VAL, (byte)AREA_TIMER, 0x00, 0x09);

    // Experimental: send an arbitrary SHORT_PARA frame. For protocol
    // probes only; vendor indices are 0..4.
    public static byte[] BuildShortParaProbe(int index, int d1, int d2) =>
        Eight(OP_SHORT_PARA,
            (byte)(index & 0xFF),
            (byte)(d1 & 0xFF),
            (byte)(d2 & 0xFF));

    private static byte[] Eight(int opcode, byte d0, byte d1, byte d2)
    {
        var o = new byte[8];
        o[0] = SYNC;
        o[1] = 0x00;
        o[2] = (byte)opcode;
        o[3] = d0;
        o[4] = d1;
        o[5] = d2;
        int crc = Crc16Xmodem(o, 0, 6);
        o[6] = (byte)((crc >> 8) & 0xFF);
        o[7] = (byte)(crc & 0xFF);
        return o;
    }

    private static readonly Random Rng = new();
    private static byte Nonce() => (byte)Rng.Next(0, 256);

    // --- Long-frame write (op 0xFF, double CRC) ------------------------

    public static byte[] BuildWriteTimerArea(IReadOnlyList<WriteTimerSlot> slots)
    {
        // Always send all 7 days; missing days default to mode=0.
        var byDay = new Dictionary<int, WriteTimerSlot>();
        foreach (var s in slots)
        {
            int d = Math.Clamp(s.DayIndex, 0, 6);
            byDay[d] = s;
        }
        var payload = new byte[TimerPayloadBytes];   // 40 bytes, trailing 5 zero
        for (int i = 0; i < 7; i++)
        {
            var s = byDay.TryGetValue(i, out var v) ? v : new WriteTimerSlot(i, 0, 0, 0, 0, 0);
            payload[5 * i + 0] = (byte)s.ModeRaw;
            payload[5 * i + 1] = (byte)s.OnHour;
            payload[5 * i + 2] = (byte)s.OnMin;
            payload[5 * i + 3] = (byte)s.OffHour;
            payload[5 * i + 4] = (byte)s.OffMin;
        }
        return BuildLongWriteFrame(AREA_TIMER, payload);
    }

    private static byte[] BuildLongWriteFrame(int areaTag, byte[] payload)
    {
        if ((payload.Length & 3) != 0)
            throw new ArgumentException($"payload must be 4-aligned, got {payload.Length}", nameof(payload));
        int total = 8 + payload.Length + 4;
        var o = new byte[total];
        o[0] = SYNC;
        o[1] = (byte)((payload.Length / 4) - 1);
        o[2] = (byte)OP_WRITE_AREA;
        o[3] = Nonce();
        // Address bytes 4..6 zero; byte 7 = area tag (vendor sends longAddr
        // reversed; [0xF5,0,0,0] → [0,0,0,0xF5]).
        o[4] = 0; o[5] = 0; o[6] = 0;
        o[7] = (byte)areaTag;
        Buffer.BlockCopy(payload, 0, o, 8, payload.Length);
        int pCrc = Crc16Xmodem(payload, 0, payload.Length);
        o[8 + payload.Length]     = (byte)((pCrc >> 8) & 0xFF);
        o[8 + payload.Length + 1] = (byte)(pCrc & 0xFF);
        int fCrc = Crc16Xmodem(o, 0, 8 + payload.Length + 2);
        o[8 + payload.Length + 2] = (byte)((fCrc >> 8) & 0xFF);
        o[8 + payload.Length + 3] = (byte)(fCrc & 0xFF);
        return o;
    }

    private const int TimerPayloadBytes = 40;

    // --- CRC-16 / XMODEM ----------------------------------------------
    // Poly 0x1021, init 0x0000, no reflection, no final XOR. Table lookup
    // mirrors the JS implementation (16-entry nibble table).

    private static readonly int[] CrcTable =
    {
        0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50A5, 0x60C6, 0x70E7,
        0x8108, 0x9129, 0xA14A, 0xB16B, 0xC18C, 0xD1AD, 0xE1CE, 0xF1EF,
    };

    private static int Crc16Xmodem(byte[] buf, int offset, int len)
    {
        int crc = 0;
        for (int i = 0; i < len; i++)
        {
            int b  = buf[offset + i] & 0xFF;
            int hi = (b >> 4) & 0x0F;
            int lo = b & 0x0F;
            crc = ((crc << 4) & 0xFFFF) ^ CrcTable[((crc >> 12) & 0x0F) ^ hi];
            crc = ((crc << 4) & 0xFFFF) ^ CrcTable[((crc >> 12) & 0x0F) ^ lo];
        }
        return crc & 0xFFFF;
    }

    public static bool ValidateCrc(byte[] bytes)
    {
        if (bytes.Length < 4) return false;
        return Crc16Xmodem(bytes, 0, bytes.Length) == 0;
    }

    // --- Incoming frames -----------------------------------------------

    private const double NotSupported = double.MinValue;

    public static HeaterTelemetry? ParseTelemetry(byte[] bytes)
    {
        if (bytes.Length < 10) return null;
        if (bytes[0] != SYNC) return null;
        if (!ValidateCrc(bytes)) return null;

        int len = bytes[1] & 0xFF;
        if (len == 0) return null;   // 8-byte short frames carry no payload

        int tag = bytes[7] & 0xFF;
        if (tag != AREA_REG_INFO) return null;

        int payloadLen = 4 * (len + 1);
        if (bytes.Length < 8 + payloadLen + 2) return null;
        var p = new byte[payloadLen];
        Buffer.BlockCopy(bytes, 8, p, 0, payloadLen);
        if (p.Length < 24) return null;

        int machineStatus  = p[0] & 0xFF;
        int runningModeNib = machineStatus & 0x0F;
        bool tempUnitF     = (machineStatus & 0x10) != 0;

        double voltage  = UleShort(p, 2) / 10.0;
        int    altitude = UleShort(p, 4);
        double ambient  = SignedTempC(p, 6);
        double housing  = SignedTempC(p, 8);
        double pumpHz   = UleShort(p, 10) / 10.0;
        double ignWatts = UleShort(p, 12) / 10.0;
        int    fanRpm   = UleShort(p, 14);
        double? intake  = NullIfNotSupported(SignedTempC(p, 16));
        double? outlet  = NullIfNotSupported(SignedTempC(p, 18));
        int    errBits  = UleShort(p, 20);

        int?    aimGear = null;
        double? targetC = null;
        if (p.Length >= 36)
        {
            aimGear = p[32] & 0xFF;
            int tval = p[33] & 0xFF;
            // Stored in the heater's current display unit (tempUnitF bit).
            // Convert F→C so app code is always Celsius.
            targetC = tempUnitF ? (tval - 32) * 5.0 / 9.0 : (double)tval;
        }

        return new HeaterTelemetry(
            OutletTempC        : outlet,
            TargetTempC        : targetC,
            FuelPumpHz         : pumpHz,
            FanRpm             : fanRpm,
            GlowPlugA          : null,
            BatteryV           : voltage,
            AmbientTempC       : ambient,
            HousingTempC       : housing,
            IntakeTempC        : intake,
            AltitudeM          : altitude,
            IgnitionWatts      : ignWatts,
            RunningMode        : RunningModeExt.FromWire(runningModeNib),
            FaultBits          : errBits,
            TempUnitFahrenheit : tempUnitF,
            AimGear            : aimGear,
            UpdatedAtMs        : DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
    }

    public static IReadOnlyList<TimerSlot>? ParseTimerInfo(byte[] bytes)
    {
        if (bytes.Length < 10) return null;
        if (bytes[0] != SYNC) return null;
        if (!ValidateCrc(bytes)) return null;
        int len = bytes[1] & 0xFF;
        if (len == 0) return null;
        if ((bytes[7] & 0xFF) != AREA_TIMER) return null;
        int payloadLen = 4 * (len + 1);
        if (bytes.Length < 8 + payloadLen + 2) return null;
        if (payloadLen < 35) return null;
        var p = new byte[payloadLen];
        Buffer.BlockCopy(bytes, 8, p, 0, payloadLen);
        var slots = new TimerSlot[7];
        for (int i = 0; i < 7; i++)
        {
            int off    = i * 5;
            int rawMod = p[off] & 0xFF;
            int modeRaw = rawMod;
            // Vendor folds raw 1/2 → display 3, ≥4 → display 4; we expose
            // the raw value as-is here so the editor can round-trip.
            slots[i] = new TimerSlot(
                DayIndex: i,
                ModeRaw : modeRaw,
                OnHour  : p[off + 1] & 0xFF,
                OnMin   : p[off + 2] & 0xFF,
                OffHour : p[off + 3] & 0xFF,
                OffMin  : p[off + 4] & 0xFF);
        }
        return slots;
    }

    private static int UleShort(byte[] p, int off) =>
        ((p[off + 1] & 0xFF) << 8) | (p[off] & 0xFF);

    private static double SignedTempC(byte[] p, int off)
    {
        int raw = UleShort(p, off);
        if (raw == 32760) return NotSupported;
        int signed = raw > 32767 ? raw - 65536 : raw;
        return signed / 10.0;
    }

    private static double? NullIfNotSupported(double v) =>
        v == NotSupported ? null : v;

    // --- Hex helpers ---------------------------------------------------

    public static string Hex(byte[] bytes) =>
        string.Join(' ', bytes.Select(b => b.ToString("X2")));

    public static byte[] FromHex(string hex)
    {
        var cleaned = new string(hex.Where(c => !char.IsWhiteSpace(c) && c != ':' && c != ',').ToArray());
        if ((cleaned.Length & 1) != 0)
            throw new ArgumentException("Hex must have an even length");
        var result = new byte[cleaned.Length / 2];
        for (int i = 0; i < result.Length; i++)
            result[i] = Convert.ToByte(cleaned.Substring(i * 2, 2), 16);
        return result;
    }
}
